overlord/devicestate/handlers_remodel.go

```diff
@@ -151,6 +178,100 @@ func (m *DeviceManager) doSetModel(t *state.Task, _ *tomb.Tomb) (err error) {
 	return nil
 }
 
+func trackedValidationSetsFromModel(st *state.State, model *asserts.Model) ([]*asserts.ValidationSet, error) {
+	currentSets, err := assertstate.TrackedEnforcedValidationSets(st)
+	if err != nil {
+		return nil, err
+	}
+
+	var fromModel []*asserts.ValidationSet
+	for _, mvs := range model.ValidationSets() {
+		for _, cvs := range currentSets.Sets() {
+			if mvs.SequenceName() == cvs.SequenceName() {
+				fromModel = append(fromModel, cvs)
+			}
+		}
+	}
+	return fromModel, nil
+}
+
+func rollBackValidationSets(st *state.State, oldSets []*asserts.ValidationSet, newSets []*asserts.ModelValidationSet, deviceCtx snapstate.DeviceContext) error {
+	for _, set := range newSets {
+		if err := assertstate.ForgetValidationSet(st, set.AccountID, set.Name, assertstate.ForgetValidationSetOpts{
+			ForceForget: true,
+		}); err != nil {
+			fmt.Println("we did not roll them back!", err)
+			return err
+		}
+	}
+
+	vSetKeys := make(map[string][]string, len(oldSets))
+	for _, vs := range oldSets {
+		sequenceName := vs.SequenceName()
+		vSetKeys[sequenceName] = vs.At().PrimaryKey
+	}
+
+	snaps, ignore, err := snapstate.InstalledSnaps(st)
+	if err != nil {
+		return err
+	}
+
+	// TODO: is this right? i think we have to ignore everything that is
+	// installed, since we won't have run the undos for the newly installed
+	// snaps yet
+	for _, sn := range snaps {
+		ignore[sn.SnapName()] = true
+	}
+
+	if err := assertstate.ApplyLocalEnforcedValidationSets(st, vSetKeys, nil, snaps, ignore); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resolveValidationSetAssertion(seq *asserts.AtSequence, db asserts.RODatabase) (asserts.Assertion, error) {
+	if seq.Sequence <= 0 {
+		hdrs, err := asserts.HeadersFromSequenceKey(seq.Type, seq.SequenceKey)
+		if err != nil {
+			return nil, err
+		}
+		return db.FindSequence(seq.Type, hdrs, -1, seq.Type.MaxSupportedFormat())
+	}
+	return seq.Resolve(db.Find)
+}
```

Comment:
We have this exact piece of code so many places, but it's hard to see a meaningful way to reuse this as many places have minor variants. (just an observation not something you need to change). But maybe something to think about at some future time.